logger/src/logger.cpp

```diff
@@ -1,18 +1,59 @@
 #include "logger.hpp"
 
-#include <filesystem>
-#include <iostream>
-#include <sstream>
-#include <string>
+#include <ctime>
 
-namespace logging
+namespace Logger
 {
-std::ostringstream logMsg;
-void logMessage(std::string message, const std::source_location& location)
+
+static std::map<Loglevel, std::string> logLevelMap = {
+    {Loglevel::INFO, "INFO"}, {Loglevel::CRITICAL, "CRITICAL"}};
+
+size_t byteCount = 0;
+
+/**
+ * @brief Generates a timestamp string
+ *
+ * Generates a timestamp string representing the current time
+ * in the format "YYYY-MM-DD HH:MM:SS"
+ *
+ * @return A string containing the formatted timestamp
+ */
+static std::string timestamp()
+{
+    time_t t = time(nullptr);
+    tm Time{};
+    gmtime_r(&t, &Time);
+    std::string timeStamp;
+    timeStamp.resize(30, '\0');
+    size_t currentTime = strftime(timeStamp.data(), timeStamp.size(),
+                                  "%Y-%m-%d %H:%M:%S", &Time);
+    timeStamp.resize(currentTime);
+    return timeStamp;
+}
+
+void logMessage(const Loglevel level, const std::string& message,
+                const std::source_location& location)
+{
+    std::string tmpString =
+        timestamp() + " [" + logLevelMap[level] + "]" + " : " +
+        std::filesystem::path(location.file_name()).filename().string() + ":" +
+        std::to_string(location.line()) + " - " + message + "\n";
+    while (byteCount + tmpString.size() > maxBufferSize)
+    {
+        byteCount = byteCount - (*circularBuffer.begin()).size();
+        circularBuffer.erase(circularBuffer.begin());
+    }
+    if (Loglevel::INFO == level)
+    {
+        circularBuffer.push_back(tmpString);
+        byteCount = byteCount + tmpString.size();
+    }
+}
+
+void logMessage(const std::string& message,
+                const std::source_location& location)
 {
-    logMsg << "File: "
-           << std::filesystem::path(location.file_name()).filename().string()
-           << " line: " << location.line() << " " << message << "\n";
+    logMessage(Loglevel::INFO, message, location);
```

Comment:
Yes, It is required to have location object in this API. 

This api is going to call logMessage() internally with the message to be appended to the buffer. Here the message is coming from a different file.
so, if the location object is not passed the "location" parameter of logMessage() considers the "logger.cpp" as the file name instead of the original caller.

Reply:
Correct, we cal close this comment.